File: quant/markowitz.py
#!/usr/bin/env python3
"""
Markowitz mean-variance portfolio optimization (min-variance, max-Sharpe, efficient frontier).

Uses pure-Python linear algebra (Cholesky / back-substitution) to stay dependency-free.
Short sales are allowed. Singular or indefinite covariance matrices are rejected;
no hidden regularization or third-party fallback changes the requested problem.

Usage:
    python3 markowitz.py --returns-csv returns.csv --rf 0.05
    # returns.csv: columns = assets, rows = daily returns (no header, comma-separated)
"""
import argparse
try:
    from ._validation import number, series, confidence_level
    from .correlation import parse_returns_csv
except ImportError:
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    from _validation import number, series, confidence_level
    from correlation import parse_returns_csv
import csv
import json
import math
import sys


def mean_vec(matrix):
    if not matrix or not matrix[0]:
        raise ValueError("matrix must contain observations and assets")
    m = len(matrix[0])
    for row in matrix:
        series(row, "matrix row")
        if len(row) != m:
            raise ValueError("matrix rows must have equal lengths")
    n = len(matrix)
    return [sum(matrix[i][j] for i in range(n)) / n for j in range(m)]


def cov_matrix(matrix):
    if not matrix or not matrix[0]:
        raise ValueError("matrix must contain observations and assets")
    m = len(matrix[0])
    for row in matrix:
        series(row, "matrix row")
        if len(row) != m:
            raise ValueError("matrix rows must have equal lengths")
    n = len(matrix)
    if n < 2:
        raise ValueError("covariance requires at least two observations")
    mu = mean_vec(matrix)
    cov = [[0.0] * m for _ in range(m)]
    for j in range(m):
        for k in range(j, m):
            s = sum((matrix[i][j] - mu[j]) * (matrix[i][k] - mu[k]) for i in range(n))
            v = s / (n - 1)
            cov[j][k] = v
            cov[k][j] = v
    return cov, mu


def cholesky(A):
    n = len(A)
    if not n or any(len(row) != n for row in A):
        raise ValueError("matrix must be non-empty and square")
    for i, row in enumerate(A):
        series(row, "matrix row")
        for j in range(i):
            if not math.isclose(A[i][j], A[j][i], rel_tol=1e-12, abs_tol=1e-15):
                raise ValueError("matrix must be symmetric")
    L = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(i + 1):
            s = sum(L[i][k] * L[j][k] for k in range(j))
            if i == j:
                v = A[i][i] - s
                if v <= 0:
                    raise ValueError("matrix must be positive definite; remove redundant assets or explicitly regularize")
                L[i][j] = math.sqrt(v)
            else:
                L[i][j] = (A[i][j] - s) / L[j][j]
    return L


def forward_sub(L, b):
    n = len(L)
    x = [0.0] * n
    for i in range(n):
        x[i] = (b[i] - sum(L[i][k] * x[k] for k in range(i))) / L[i][i]
    return x


def backward_sub(U, b):
    n = len(U)
    x = [0.0] * n
    for i in range(n - 1, -1, -1):
        x[i] = (b[i] - sum(U[i][k] * x[k] for k in range(i + 1, n))) / U[i][i]
    return x


def solve_linear(A, b):
    """Solve Ax=b via Cholesky. A must be positive definite."""
    series(b, "right-hand side")
    if len(b) != len(A):
        raise ValueError("right-hand side length must match matrix")
    L = cholesky(A)
    y = forward_sub(L, b)
    Lt = [[L[j][i] for j in range(len(L))] for i in range(len(L))]
    x = backward_sub(Lt, y)
    return x


def min_variance_portfolio(cov):
    """Global minimum variance: w = cov^-1 * 1 / (1' * cov^-1 * 1)"""
    n = len(cov)
    ones = [1.0] * n
    inv_cov_ones = solve_linear(cov, ones)
    denom = sum(inv_cov_ones)
    weights = [v / denom for v in inv_cov_ones]
    return weights


def max_sharpe_portfolio(cov, excess_returns):
    """
    Max-Sharpe tangency: w proportional to cov^-1 * excess_returns, normalized to sum=1.
    """
    inv_cov_r = solve_linear(cov, excess_returns)
    denom = sum(inv_cov_r)
    if denom <= 1e-12:
        return None
    return [v / denom for v in inv_cov_r]


def portfolio_stats(weights, mu, cov, rf=0.0, annualize=252):
    ret = sum(w * m for w, m in zip(weights, mu))
    var = 0.0
    n = len(weights)
    for i in range(n):
        for j in range(n):
            var += weights[i] * weights[j] * cov[i][j]
    vol = math.sqrt(max(0, var))
    sharpe = (ret * annualize - rf) / (vol * math.sqrt(annualize)) if vol > 0 else 0.0
    return {
        "expected_return_ann_pct": round(ret * annualize * 100, 3),
        "volatility_ann_pct": round(vol * math.sqrt(annualize) * 100, 3),
        "sharpe": round(sharpe, 3),
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--returns-csv", required=True)
    ap.add_argument("--asset-names", help="comma-separated names (optional, else col0..colN)")
    ap.add_argument("--rf", type=float, default=0.05)
    ap.add_argument("--annualize", type=int, default=252)
    args = ap.parse_args()

    with open(args.returns_csv) as f:
        returns = parse_returns_csv(f, minimum_rows=0, minimum_assets=1)

    if not returns:
        print(json.dumps({"error": "no numeric data found"}))
        sys.exit(1)

    n_assets = len(returns[0])
    names = args.asset_names.split(",") if args.asset_names else [f"asset_{i}" for i in range(n_assets)]
    if len(names) != n_assets or len(set(names)) != n_assets or any(not name.strip() for name in names):
        raise ValueError("asset names must be unique, non-empty and match the data columns")
    number(args.rf, "rf")
    if args.annualize <= 0:
        raise ValueError("annualize must be positive")
    cov, mu = cov_matrix(returns)
    periodic_rf = args.rf / args.annualize
    excess = [m - periodic_rf for m in mu]

    out = {"n_assets": n_assets, "n_observations": len(returns), "asset_names": names}

    mv_w = min_variance_portfolio(cov)
    out["min_variance_portfolio"] = {
        "weights": {n: round(w, 4) for n, w in zip(names, mv_w)},
        **portfolio_stats(mv_w, mu, cov, args.rf, args.annualize),
    }

    ms_w = max_sharpe_portfolio(cov, excess)
    if ms_w:
        out["max_sharpe_portfolio"] = {
            "weights": {n: round(w, 4) for n, w in zip(names, ms_w)},
            **portfolio_stats(ms_w, mu, cov, args.rf, args.annualize),
        }

    # equal-weight benchmark
    ew_w = [1 / n_assets] * n_assets
    out["equal_weight_benchmark"] = {
        "weights": {n: round(1 / n_assets, 4) for n in names},
        **portfolio_stats(ew_w, mu, cov, args.rf, args.annualize),
    }

    print(json.dumps(out, indent=2))


if __name__ == "__main__":
    main()

File: quant-jvm/src/main/kotlin/org/maxmoran/quant/Markowitz.kt
package org.maxmoran.quant

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import java.io.File
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt
import kotlin.system.exitProcess

private val PRETTY_JSON = Json { prettyPrint = true }

/**
 * Deterministic port of `quant/markowitz.py` — min-variance, max-Sharpe, and equal-weight
 * portfolios from a returns matrix.
 *
 * parity-contract.md §1.4 applies: the reference solves cov * x = b with a hand-rolled
 * Cholesky factorization plus forward/backward substitution (pure Python, no numpy in the
 * code path), so this port replicates that exact arithmetic — same loops, same summation
 * order, same 1e-10 pivot regularization — rather than delegating to a library solver.
 * Raw solver parity is asserted at the §1.4 tolerance (1e-6 per entry).
 */

/** Column means of a row-major returns matrix; mirrors `mean_vec`. */
fun meanVec(matrix: List<List<Double>>): List<Double> {
    val n = matrix.size
    val m = matrix[0].size
    return List(m) { j -> matrix.sumOf { it[j] } / n }
}

/** Sample (ddof=1) covariance matrix plus the mean vector; mirrors `cov_matrix`. */
fun covMatrix(matrix: List<List<Double>>): Pair<List<List<Double>>, List<Double>> {
    val n = matrix.size
    val m = matrix[0].size
    val mu = meanVec(matrix)
    val cov = Array(m) { DoubleArray(m) }
    for (j in 0 until m) {
        for (k in j until m) {
            var s = 0.0
            for (i in 0 until n) s += (matrix[i][j] - mu[j]) * (matrix[i][k] - mu[k])
            val v = s / (n - 1)
            cov[j][k] = v
            cov[k][j] = v
        }
    }
    return cov.map { it.toList() } to mu
}

/** Cholesky factor with the reference's 1e-10 regularization of non-positive pivots. */
fun cholesky(a: List<List<Double>>): List<List<Double>> {
    val n = a.size
    val l = Array(n) { DoubleArray(n) }
    for (i in 0 until n) {
        for (j in 0..i) {
            var s = 0.0
            for (k in 0 until j) s += l[i][k] * l[j][k]
            if (i == j) {
                var v = a[i][i] - s
                if (v <= 0.0) v = 1e-10
                l[i][j] = sqrt(v)
            } else {
                l[i][j] = (a[i][j] - s) / l[j][j]
            }
        }
    }
    return l.map { it.toList() }
}

private fun forwardSub(l: List<List<Double>>, b: List<Double>): List<Double> {
    val n = l.size
    val x = DoubleArray(n)
    for (i in 0 until n) {
        var s = 0.0
        for (k in 0 until i) s += l[i][k] * x[k]
        x[i] = (b[i] - s) / l[i][i]
    }
    return x.toList()
}

private fun backwardSub(u: List<List<Double>>, b: List<Double>): List<Double> {
    val n = u.size
    val x = DoubleArray(n)
    for (i in n - 1 downTo 0) {
        var s = 0.0
        for (k in i + 1 until n) s += u[i][k] * x[k]
        x[i] = (b[i] - s) / u[i][i]
    }
    return x.toList()
}

/** Solve Ax = b via Cholesky, matching `solve_linear`'s factor-transpose-substitute path. */
fun solveLinear(a: List<List<Double>>, b: List<Double>): List<Double> {
    val l = cholesky(a)
    val y = forwardSub(l, b)
    val lt = List(l.size) { i -> List(l.size) { j -> l[j][i] } }
    return backwardSub(lt, y)
}

/** Global minimum variance weights: cov^-1 * 1, normalized to sum to one. */
fun minVariancePortfolio(cov: List<List<Double>>): List<Double> {
    val ones = List(cov.size) { 1.0 }
    val invCovOnes = solveLinear(cov, ones)
    val denom = invCovOnes.sum()
    return invCovOnes.map { it / denom }
}

/** Max-Sharpe tangency weights, or null when the normalizer is degenerate (|sum| < 1e-12). */
fun maxSharpePortfolio(cov: List<List<Double>>, excessReturns: List<Double>): List<Double>? {
    val invCovR = solveLinear(cov, excessReturns)
    val denom = invCovR.sum()
    if (abs(denom) < 1e-12) return null
    return invCovR.map { it / denom }
}

/** Annualized return / volatility / Sharpe block, rounded like `portfolio_stats`. */
fun portfolioStats(
    weights: List<Double>,
    mu: List<Double>,
    cov: List<List<Double>>,
    rf: Double = 0.0,
    annualize: Int = 252,
): JsonObject {
    var ret = 0.0
    for (i in weights.indices) ret += weights[i] * mu[i]
    var variance = 0.0
    val n = weights.size
    for (i in 0 until n) {
        for (j in 0 until n) {
            variance += weights[i] * weights[j] * cov[i][j]
        }
    }
    val vol = sqrt(max(0.0, variance))
    val annualizeSqrt = sqrt(annualize.toDouble())
    val sharpe = if (vol > 0.0) (ret * annualize - rf) / (vol * annualizeSqrt) else 0.0
    return buildJsonObject {
        put("expected_return_ann_pct", JsonPrimitive(round3(ret * annualize * 100.0)))
        put("volatility_ann_pct", JsonPrimitive(round3(vol * annualizeSqrt * 100.0)))
        put("sharpe", JsonPrimitive(round3(sharpe)))
    }
}

/** Public rounded JSON value contract, field-for-field with `quant/markowitz.py`. */
fun markowitzOutput(
    returns: List<List<Double>>,
    assetNames: List<String>?,
    rf: Double = 0.05,
    annualize: Int = 252,
): JsonObject {
    if (returns.isEmpty()) throw IllegalArgumentException("no numeric data found")
    val nAssets = returns[0].size
    val names = assetNames ?: List(nAssets) { "asset_$it" }
    val (cov, mu) = covMatrix(returns)
    val periodicRf = rf / annualize
    val excess = mu.map { it - periodicRf }

    return buildJsonObject {
        put("n_assets", JsonPrimitive(nAssets))
        put("n_observations", JsonPrimitive(returns.size))
        put("asset_names", buildJsonArray { names.forEach { add(JsonPrimitive(it)) } })

        val minVarianceWeights = minVariancePortfolio(cov)
        put("min_variance_portfolio", portfolioBlock(names, minVarianceWeights, mu, cov, rf, annualize))

        // Reference omits the key entirely when the tangency normalizer is degenerate.
        maxSharpePortfolio(cov, excess)?.let { maxSharpeWeights ->
            put("max_sharpe_portfolio", portfolioBlock(names, maxSharpeWeights, mu, cov, rf, annualize))
        }

        val equalWeights = List(nAssets) { 1.0 / nAssets }
        put("equal_weight_benchmark", buildJsonObject {
            // Reference iterates ALL names here (not zip), so surplus names appear too.
            put("weights", buildJsonObject {
                names.forEach { put(it, JsonPrimitive(round4(1.0 / nAssets))) }
            })
            portfolioStats(equalWeights, mu, cov, rf, annualize).forEach { (key, value) -> put(key, value) }
        })
    }
}

// Python's zip(names, weights) truncates to the shorter sequence; duplicates overwrite.
private fun portfolioBlock(
    names: List<String>,
    weights: List<Double>,
    mu: List<Double>,
    cov: List<List<Double>>,
    rf: Double,
    annualize: Int,
): JsonObject = buildJsonObject {
    put("weights", buildJsonObject {
        names.zip(weights).forEach { (name, w) -> put(name, JsonPrimitive(round4(w))) }
    })
    portfolioStats(weights, mu, cov, rf, annualize).forEach { (key, value) -> put(key, value) }
}

/** Testable result of evaluating the Markowitz CLI contract. */
data class MarkowitzCliResult(val exitCode: Int, val output: String)

/** Parse input and produce CLI output without terminating the process. */
fun evaluateMarkowitz(args: Array<String>): MarkowitzCliResult {
    return try {
        var csvPath: String? = null
        var assetNames: String? = null
        var rf = 0.05
        var annualize = 252
        var index = 0

        fun optionValue(option: String): String {
            if (index + 1 >= args.size || args[index + 1].startsWith("--")) {
                throw IllegalArgumentException("$option requires a value")
            }
            index += 1
            return args[index]
        }

        while (index < args.size) {
            when (val arg = args[index]) {
                "--returns-csv" -> csvPath = optionValue(arg)
                "--asset-names" -> assetNames = optionValue(arg)
                "--rf" -> rf = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--rf must be a number")
                "--annualize" -> annualize = optionValue(arg).toIntOrNull()
                    ?: throw IllegalArgumentException("--annualize must be an integer")
                else -> throw IllegalArgumentException("unknown argument: $arg")
            }
            index += 1
        }

        val path = csvPath ?: throw IllegalArgumentException("need --returns-csv")
        val returns = parseReturnsCsv(File(path).readText())
        if (returns.isEmpty()) throw IllegalArgumentException("no numeric data found")
        val names = assetNames?.split(",")
        MarkowitzCliResult(
            exitCode = 0,
            output = PRETTY_JSON.encodeToString(
                JsonObject.serializer(),
                markowitzOutput(returns, names, rf, annualize),
            ),
        )
    } catch (exception: Exception) {
        val message = exception.message?.lineSequence()?.firstOrNull()?.take(240) ?: "invalid input"
        MarkowitzCliResult(
            exitCode = 1,
            output = buildJsonObject { put("error", JsonPrimitive(message)) }.toString(),
        )
    }
}

/** Module CLI runner — CSV in / JSON out, matching the Python entrypoint. */
fun runMarkowitz(args: Array<String>) {
    val result = evaluateMarkowitz(args)
    println(result.output)
    if (result.exitCode != 0) exitProcess(result.exitCode)
}
